source/lexbor/dom/interfaces/attr.c

```diff
@@ -82,54 +82,73 @@ lxb_dom_attr_set_name(lxb_dom_attr_t *attr, const lxb_char_t *name,
 }
 
 lxb_status_t
-lxb_dom_attr_set_name_ns(lxb_dom_attr_t *attr, const lxb_char_t *link,
-                         size_t link_length, const lxb_char_t *name,
-                         size_t name_length, bool to_lowercase)
+lxb_dom_attr_set_name_ns(lxb_dom_attr_t *attr,
+                         const lxb_char_t *local_name, size_t local_name_len,
+                         const lxb_char_t *ns_link, size_t ns_len,
+                         const lxb_char_t *prefix, size_t prefix_len,
+                         bool to_lowercase)
 {
-    size_t length;
-    lxb_char_t *p;
+    size_t qualified_name_len = local_name_len;
+    lxb_char_t *qualified_name = (lxb_char_t *) local_name;
     const lxb_ns_data_t *ns_data;
     lxb_dom_attr_data_t *data;
     lxb_dom_document_t *doc = lxb_dom_interface_node(attr)->owner_document;
 
-    ns_data = lxb_ns_append(doc->ns, link, link_length);
-    if (attr->node.ns == LXB_NS__UNDEF) {
+    if (ns_link != NULL && ns_len != 0) {
+        ns_data = lxb_ns_append(doc->ns, ns_link, ns_len);
+    }
+    else {
+        ns_data = lxb_ns_data_by_id(doc->ns, LXB_NS__UNDEF);
+    }
+
+    if (ns_data == NULL) {
         return LXB_STATUS_ERROR;
     }
 
     attr->node.ns = ns_data->ns_id;
 
     /* TODO: append check https://www.w3.org/TR/xml/#NT-Name */
 
-    p = (lxb_char_t *) memchr(name, ':', name_length);
-    if (p == NULL) {
-        return lxb_dom_attr_set_name(attr, name, name_length, to_lowercase);
-    }
+    if (prefix != NULL && prefix_len != 0) {
+        qualified_name = lexbor_malloc(prefix_len + local_name_len + 2);
```

Comment:
@Azq2 

I don't like the temporary allocation of memory here.

We need to change 
```C
lxb_dom_attr_data_t *
lxb_dom_attr_qualified_name_append(lexbor_hash_t *hash, const lxb_char_t *name,
                                   size_t length)
```
to
```C
lxb_dom_attr_data_t *
lxb_dom_attr_qualified_name_append(lexbor_hash_t *hash, const lxb_char_t *name,
                                   size_t length, bool copy)
```

And use in here `lexbor_mraw_alloc(doc->attrs->mraw, prefix_len + local_name_len + 2);`
And create `lexbor_hash_insert_raw_wo_copy` for use in `lxb_dom_attr_qualified_name_append`:
```C
if (copy) {
    data = lexbor_hash_insert(hash, lexbor_hash_insert_raw, name, length);
}
else {
    data = lexbor_hash_insert(hash, lexbor_hash_insert_raw_wo_copy, name, length);
}

if ((lxb_dom_attr_id_t) data <= LXB_DOM_ATTR__LAST_ENTRY) {
```

Reply:
I copied the same code from https://github.com/lexbor/lexbor/blob/master/source/lexbor/dom/interfaces/element.c#L134 =)

Okay. But current hash implementation without `lexbor_hash_insert_raw_wo_copy`:
```error: ‘lexbor_hash_insert_raw_wo_copy’ undeclared (first use in this function);```

I try implement it, but have some problems. 
Allocated mraw needed only for `length <= LEXBOR_HASH_SHORT_SIZE`: https://github.com/lexbor/lexbor/blob/master/source/lexbor/core/hash.c#L405

Because of this, for short values ​​there will be either a leak or a layer violation (freeing mraw in lexbor_hash_wo_copy if it not claimed)

Reply:
@Azq2 

> I copied the same code from https://github.com/lexbor/lexbor/blob/master/source/lexbor/dom/interfaces/element.c#L134 =)

My mistake, it will be necessary to fix this in the future.
Okay, let's leave it for now.

In general, it’s better to create at least some tests for the new functionality.
Can we add a couple of tests? To `test/lexbor/html/dom_element.c`, for example.